Restructure GDT. Add sysenter support

sktt opened the pull request:
Based on #284. Request for comments on this. 

In particular the GDT bit that had to be reworked a bit. Also, whether and how we should give the option to use sysenter instead of syscall.

sktt committed:
segment: restructure and document GDT selectors

The GDT entries needed to be re-organized to support sysexit. While
doing so, better document why they exist in the first place, and their
particular order.

Rename __KERN_{CS,DS}16 -> __RMODE_{CS,DS}16 to clearly indicate that
these entries only exist in the real mode GDT.

Omit entries that are unused in i386 or unused for a particular gdt.

Signed-off-by: Johannes Wikner <johannes.wikner@gmail.com>

sktt committed:
usermode: adjust MSR_STAR.user_cs

user_cs should point to USER_CS32 so that correct %cs and %ss are set
upon sysret and sysretq.

This fixes a subtle bug which where branches using %cs (e.g., intra-mode
iret) would crash after a syscall.

Signed-off-by: Johannes Wikner <johannes.wikner@gmail.com>

sktt committed:
entry, usermode: add sysenter handler

Introduces the necessary MSRs, and sets them to handle sysenter and
sysexits. Add macro to choose approriate instruction depending on build.

When adding compatibility mode support, we need to use either sysexit or
sysexitq depending on what usermode is executing. Same for sysret and
sysretq.

Signed-off-by: Johannes Wikner <johannes.wikner@gmail.com>

sktt committed:
usermode: add sysenter and syscall_mode selection

To give usermode the option to choose sysenter instead of syscall, we
expose syscall_mode(enum syscall_mode), which is by default using
syscall.

Signed-off-by: Johannes Wikner <johannes.wikner@gmail.com>

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -208,6 +208,26 @@ ENTRY(syscall_handler_entry)
 
+ENTRY(sysenter_handler_entry)
+    push %_ASM_CX
+    push %_ASM_DX
```
As mentioned in https://github.com/KernelTestFramework/ktf/pull/284#discussion_r1305451998, all callee-clobbered registers should be saved. `syscall_exit` also expects `flags` to be the top stack word, which SYSENTER doesn't provide -- it needs to be pushed manually.

Also, SYSENTER enters the kernel with user FLAGS, only IF and VM are forced to zero. So we might want to sanitize it (e.g. unset TF, AC) before executing further kernel code?

wipawel replied:
The `flags` are pushed onto kernel stack by `enter_usermode()` when the task is created, so this aspect should be fine AFAICS.

Right, we have to sanitize the flags for example with the mask used for SYSCALL's MSR_FMASK.

wipawel replied:
Please add `MASK_USER_FLAGS` after `syscall_from_usermode`

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -208,6 +208,26 @@ ENTRY(syscall_handler_entry)
+
+    syscall_to_usermode
+    pop %_ASM_DX
+    pop %_ASM_CX
```
Should we sanity check DX and CX somehow, so userland can't make them point to arbitrary addresses, e.g. into the non-cannonical address space?

wipawel replied:
If we want to do this, the correct range for usermode would be: `VIRT_USER_BASE` - `0x00007FFFFFFFFFFF`

sktt replied:
I think it's not too bad to just let it crash if userland provides bad values. I don't know if for example masking them to valid userland addresses would make such a crash less cryptic..

sktt replied:
So then, it would be to compare and raise and error already here... Since will no longer expose the low-level sysenter(), maybe there's no need to care?

minipli-oss replied:
Well, it matters for *how* it crashes. When RCX or RDX are non-canonical, SYSEXIT will #GP while still in kernel mode instead of, e.g., #PF back in userland when RIP is canonical but not mapped. Slightly different semantics, maybe worth to be aware of?
Not to mention SYSEXIT will #UD on AMD when executed in long mode.

sktt replied:
oh yes then it makes absolute sense. I assumed the crash would always happen in user mode.
The easy way out is to just mask it.. Is that enough or better terminate the usermode task somehow?

sysexit: yes, and sysenter too.. Do we want to warn AMD-users about this?

sktt replied:
maybe these can be a separate issue. There's also the issue with `syscall`  from `0x7fffffffffff` which consequentially crashes the kernel (on intel i think) on a `sysret`. Maybe those issues can be handles separately? OTOH, maybe this kind of stuff is exactly what we want to easily try out on KTF?

wipawel replied:
I agree, let's handle it separately.

minipli-oss commented on common/usermode.c:
```diff
@@ -124,13 +124,43 @@ static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, lo
+    asm volatile (
+        SAVE_CALLEE_SAVED_REGS()
+        "mov %%" STR(_ASM_SP) ", %%" STR(_ASM_CX) "\n"
+        "mov $1f, %%" STR(_ASM_DX) "\n"
```
Please use `lea 1f(%rip), %rcx` for x86-64 to avoid the TEXTREL entry.

minipli-oss commented on common/usermode.c:
```diff
@@ -124,13 +124,43 @@ static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, lo
+    register long return_code asm(STR(_ASM_AX));
+    register long _arg4 asm("r8") = arg4;
+    register long _arg5 asm("r9") = arg5;
+    register long _arg3 asm("r10") = arg3;
```
Is it an issue that these can't be used in legacy mode, i.e. in 32 bit mode?

wipawel replied:
It will be one day... let's maybe add a BUILD_BUG_ON() assert here for 32bit builds?

minipli-oss replied:
Not needed, I guess, as the compiler will already complain that it knows nothing about these registers ;)

minipli-oss commented on common/usermode.c:
```diff
@@ -124,13 +124,43 @@ static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, lo
+#define sysenter3(nr, a1, a2, a3)         sysenter((nr), (a1), (a2), (a3), 0, 0)
+#define sysenter4(nr, a1, a2, a3, a4)     sysenter((nr), (a1), (a2), (a3), (a4), 0)
+#define sysenter5(nr, a1, a2, a3, a4, a5) sysenter((nr), (a1), (a2), (a3), (a4), (a5))
+
```
Hmm, do we need those? Using `SYSCALL` or `SYSENTER` (or `INT $vec` if we support this someday) is an "implementation detail" of the syscall ABI and not meant to be used directly -- especially as the various instructions have all their specialties to handle.

I think an generic `syscall(...)` API is sufficient with the implementation details all handled within this function. However, to support testing different instructions, a "syscall mode switch" should be provided, e.g. something like this:

```c
enum syscall_mode {
  SYSCALL_MODE_SYSCALL,     // use SYSCALL
  SYSCALL_MODE_SYSENTER,    // use SYSENTER
  SYSCALL_MODE_INT,         // TODO: use INT $SYSVEC
};

bool syscall_mode(enum syscall_mode);
```

wipawel replied:
This is good idea!

minipli-oss commented on include/usermode.h:
```diff
@@ -53,6 +53,12 @@ extern void __user_text printf(const char *fmt, ...);
+
+extern long __user_text sysenter(long syscall_nr, long arg1, long arg2, long arg3,
+                                        long arg4, long arg5);
+
```
See https://github.com/KernelTestFramework/ktf/pull/290/commits/2d5d2d59729b0ab0db0bd02a0b82a323c3b2d587#r1307195325, this low level API shouldn't be exported, imho, but a _mode switching API_ instead.

minipli-oss reviewed (changes requested):
Thanks a lot for doing this cleanup! The comments in segment.h are especially helpful, I think!

Can you please rebase your branch, as some of the depending commits got merged?

sktt committed:
usermode: support int 80h syscalls

From usermode's perspective it works the same as syscall.

Signed-off-by: Johannes Wikner <johannes.wikner@gmail.com>
Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel commented on arch/x86/entry.S:
```diff
@@ -211,6 +211,7 @@ ENTRY(syscall_handler_entry)
 
     syscall_to_usermode
     RESTORE_CALLEE_SAVED_REGS
+
```
This commit is no longer needed

wipawel commented on common/usermode.c:
```diff
@@ -135,6 +135,54 @@ static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, lo
+        return _sysenter(syscall_nr, arg1, arg2, arg3, arg4, arg5);
+    case SYSCALL_MODE_INT:
+        // unimplemented
+        UNREACHABLE();
```
please add `default:` and then the next `UNREACHABLE()` is not needed I think

wipawel commented on include/usermode.h:
```diff
@@ -52,6 +58,7 @@ extern void __user_text exit(unsigned long exit_code);
 extern void __user_text printf(const char *fmt, ...);
 extern void *__user_text mmap(void *va, unsigned long order);
 extern void __user_text munmap(void *va, unsigned long order);
+bool __user_text syscall_mode(enum syscall_mode);
```
nit: please add extern

wipawel commented on include/usermode.h:
```diff
@@ -34,6 +34,12 @@
+  SYSCALL_MODE_SYSCALL,     // use SYSCALL
+  SYSCALL_MODE_SYSENTER,    // use SYSENTER
+  SYSCALL_MODE_INT,         // TODO: use INT $SYSVEC
+};
```
I would add a typedef here and use it everywhere

wipawel commented on common/usermode.c:
```diff
@@ -135,6 +135,54 @@ static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, lo
+static __user_bss enum syscall_mode sc_mode = SYSCALL_MODE_SYSCALL;
+
+bool __user_text syscall_mode(enum syscall_mode mode) {
+    if (mode < SYSCALL_MODE_SYSCALL || mode > SYSCALL_MODE_INT) {
```
enum is `unsigned int`, so for this check to work the SYSCALL_MODE_SYSCALL should start at 1

minipli-oss replied:
We shouldn't change SYSCALL_MODE_SYSCALL but leave the code as-is -- kind of defensive programming.

sktt replied:
only checking upper bound instead. good so?

wipawel commented on arch/x86/entry.S:
```diff
@@ -173,9 +181,15 @@ ENTRY(enter_usermode)
     PUSHF
 
     /* Save stack pointer onto per-cpu */
-    mov %_ASM_DX, %gs:(usermode_private)
+    mov %_ASM_SP, %gs:usermode_private
 
-    syscall_to_usermode
+    /* provide sysenter the kernel stack to be used */
+    push %_ASM_DX
+    WRITE_MSR $MSR_SYSENTER_ESP, %_ASM_SP
+    pop %_ASM_DX
+
+    /* Move to user stack */
+    mov %_ASM_DX, %_ASM_SP
```
We should keep the previous logic and order here. Just add the PUSH+WRMSR+POP after the PUSHF

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -173,9 +181,15 @@ ENTRY(enter_usermode)
+    /* provide sysenter the kernel stack to be used */
+    push %_ASM_DX
+    WRITE_MSR $MSR_SYSENTER_ESP, %_ASM_SP
+    pop %_ASM_DX
```
Do we really want to reprogram this MSR for each task? Why can't this simply be `percpu->tss.rsp0` and programmed just once in ?

wipawel replied:
This is a good suggestion. Just make sure the `tss.rsp0` stack does not end up in the `usermode_private` var, as this is a placeholder for either kernel (main one) or user stack

minipli-oss commented on common/usermode.c:
```diff
@@ -91,6 +97,12 @@ static void init_syscall(void) {
+static void init_sysenter(void) {
+    wrmsr(MSR_SYSENTER_CS, _ul(__KERN_CS));
+    wrmsr(MSR_SYSENTER_ESP, _ul(0)); /* configure the stack on enter_usermode instead */
+    wrmsr(MSR_SYSENTER_EIP, _ul(&sysenter_handler_entry));
```
Please don't, just use `percpu->tss.rsp0` here.

minipli-oss commented on common/usermode.c:
```diff
@@ -125,6 +158,52 @@ static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, lo
+    return return_code;
+}
+
+static __user_bss enum syscall_mode sc_mode = SYSCALL_MODE_SYSCALL;
```
`__user_data`, as it gets initialized (admitted, the value is `0`, but lets not be overly cleaver here).

minipli-oss commented on common/usermode.c:
```diff
@@ -105,10 +117,31 @@ void init_usermode(percpu_t *percpu) {
     vmap_user_4k(usermode_helpers, virt_to_mfn(usermode_helpers), L1_PROT);
 
     init_syscall();
+    init_sysenter();
```
Should we test if the CPU actually supports SEP or simply assume we're running on a non-ancient CPU?

wipawel replied:
We have this ancient issue #13, which is supposed to add proper enumerations of CPU flags and features and then use it accordingly. I would handle this there (when finally someone decides to implement it ;-))

wipawel commented on include/arch/x86/segment.h:
```diff
@@ -29,47 +30,87 @@
 #define __USER_CS32 ((GDT_USER_CS32 << 3) | 3)
 #define __USER_DS32 ((GDT_USER_DS32 << 3) | 3)
 #define __USER_CS64 ((GDT_USER_CS64 << 3) | 3)
-#define __USER_DS64 ((GDT_USER_DS64 << 3) | 3)
+#define __USER_DS64 ((GDT_USER_DS << 3) | 3)
```
nit: do we need #if defined(__x86_64__) here?

wipawel commented on include/arch/x86/segment.h:
```diff
@@ -29,47 +30,87 @@
-#define GDT_PERCPU 0xb
+/* Boot GDT */
+/* TSS is not strictly required for boot, but useful to handle early-stage faults. */
+#define GDT_BOOT_TSS 4 /* 2 entries in 64-bit */
```
This could be number 3 in 32bit I think

wipawel commented on include/arch/x86/segment.h:
```diff
@@ -29,47 +30,87 @@
-#define NR_RMODE_GDT_ENTRIES 9
-#define NR_GDT_ENTRIES       12
+/* Real mode GDT */
+#define GDT_RMODE_CS16 4
```
This could be number 3 in 32bit I think

wipawel commented on common/usermode.c:
```diff
@@ -27,12 +27,15 @@
 extern char interrupt_handlers[], end_interrupt_handlers[];
 extern char usermode_helpers[], end_usermode_helpers[];
 
+extern void int80_handler_entry(void);
+
```
please add this to `include/usermode.h` along the other entry handlers

wipawel commented on common/usermode.c:
```diff
@@ -111,6 +114,26 @@ void init_usermode(percpu_t *percpu) {
+        "int $" STR(SYSCALL_INT) "\n"
+        : "=a"(return_code)
+        : "0"(syscall_nr), "S"(arg1), "d"(arg2), "D" (arg3), "r"(_arg4), "r"(_arg5)
+        : STR(_ASM_CX), "r11"
```
this clobbering is not needed

minipli-oss replied:
You're right, but looking at the syscall handlers I can see that we don't guarantee that (even `syscall_handler_entry` is wrong in this regard!).

By making use of `SAVE_CALLEE_SAVED_REGS` we actually save the _wrong_ set of registers -- the ones already preserved by called C functions. The callee clobbered ones will be, well, clobbered.

It should either be `SAVE_ALL_REGS`, for simplicity, or we need to introduce a `SAVE_CALLEE_CLOBBERED_REGS` macro that takes care of preserving the callee clobbered registers.

wipawel replied:
Oh yes, I agree. I would go with the `SAVE_ALL_REGS` for the simplicity in all the handlers.

minipli-oss replied:
@sktt, as you're touching the code already, can you fix that as well? For the sysenter and int80 handlers added in this series directly, and for the syscall handler with a separate commit?

wipawel commented on arch/x86/entry.S:
```diff
@@ -225,6 +225,7 @@ END_FUNC(syscall_handler_entry)
 ENTRY(sysenter_handler_entry)
```
please squash this whole commit with https://github.com/KernelTestFramework/ktf/pull/290/commits/b7a0639636a5413363e4250f2aa9802dcc91347c

sktt replied:
this change is already https://github.com/KernelTestFramework/ktf/commit/b7a0639636a5413363e4250f2aa9802dcc91347c ?

wipawel commented on arch/x86/entry.S:
```diff
@@ -180,9 +180,10 @@ ENTRY(enter_usermode)
     SAVE_ALL_REGS
```
Please squash this whole commit with https://github.com/KernelTestFramework/ktf/pull/290/commits/b7a0639636a5413363e4250f2aa9802dcc91347c

sktt committed:
entry: syscall_handler_entry save/restore all regs

This allows us to use syscall from inline asm without clobbering all
callee-clobbered registers. Saving/restoring all registers also prevents
"leaking" kernel register values into userland.

Signed-off-by: Johannes Wikner <johannes.wikner@gmail.com>

minipli-oss reviewed (approved):
Thanks a lot, Johannes!
